### ISSUE TITLE
intermittend 3D rendering issues

### ISSUE THREAD

**Issue by wiebkejahr** [napari/napari#817]
I continue to run into a bug when switching into the 3D mode of napari viewer.
It's not yet fully reproducible; I will continue testing but might need help as to what I should test.

Workflow:
I used iPython and the following code to load the data
```
In [1]: %gui qt5
In [2]: import numpy as np
In [3]: import matplotlib.pyplot as plt
In [4]: from dask_image.imread import imread
In [5]: import napari
In [6]: stack = imread("path/to/data")
In [7]: viewer = napari.view_image(stack, contrast_limits=[0,5000],scale=[5,1,1], is_pyramid=False, colormap='magma')
```
The bug seems to occur mostly when I go to the center of the stack and then enter 3D view relatively quickly after opening napari.

I read through #576 but this one seems to be different.
vispy version 0.6.4
napari version 0.2.6

Here's the error message:
```
WARNING: Error drawing visual <Volume at 0x126ff5a2608>
Traceback (most recent call last):
  File "Z:\Python\envs\wjahr\playground\lib\site-packages\OpenGL\latebind.py", line 41, in __call__
    return self._finalCall( *args, **named )
TypeError: 'NoneType' object is not callable

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "Z:\Python\envs\wjahr\playground\lib\site-packages\vispy\app\backends\_qt.py", line 825, in paintGL
    self._vispy_canvas.events.draw(region=None)
  File "Z:\Python\envs\wjahr\playground\lib\site-packages\vispy\util\event.py", line 455, in __call__
    self._invoke_callback(cb, event)
  File "Z:\Python\envs\wjahr\playground\lib\site-packages\vispy\util\event.py", line 475, in _invoke_callback
    self, cb_event=(cb, event))
  File "Z:\Python\envs\wjahr\playground\lib\site-packages\vispy\util\event.py", line 471, in _invoke_callback
    cb(event)
  File "Z:\Python\envs\wjahr\playground\lib\site-packages\vispy\scene\canvas.py", line 217, in on_draw
    self._draw_scene()
  File "Z:\Python\envs\wjahr\playground\lib\site-packages\vispy\scene\canvas.py", line 266, in _draw_scene
    self.draw_visual(self.scene)
  File "Z:\Python\envs\wjahr\playground\lib\site-packages\vispy\scene\canvas.py", line 304, in draw_visual
    node.draw()
  File "Z:\Python\envs\wjahr\playground\lib\site-packages\vispy\scene\visuals.py", line 99, in draw
    self._visual_superclass.draw(self)
  File "Z:\Python\envs\wjahr\playground\lib\site-packages\vispy\visuals\visual.py", line 443, in draw
    self._vshare.index_buffer)
  File "Z:\Python\envs\wjahr\playground\lib\site-packages\vispy\visuals\shaders\program.py", line 101, in draw
    Program.draw(self, *args, **kwargs)
  File "Z:\Python\envs\wjahr\playground\lib\site-packages\vispy\gloo\program.py", line 533, in draw
    canvas.context.flush_commands()
  File "Z:\Python\envs\wjahr\playground\lib\site-packages\vispy\gloo\context.py", line 176, in flush_commands
    self.glir.flush(self.shared.parser)
  File "Z:\Python\envs\wjahr\playground\lib\site-packages\vispy\gloo\glir.py", line 572, in flush
    self._shared.flush(parser)
  File "Z:\Python\envs\wjahr\playground\lib\site-packages\vispy\gloo\glir.py", line 494, in flush
    parser.parse(self._filter(self.clear(), parser))
  File "Z:\Python\envs\wjahr\playground\lib\site-packages\vispy\gloo\glir.py", line 819, in parse
    self._parse(command)
  File "Z:\Python\envs\wjahr\playground\lib\site-packages\vispy\gloo\glir.py", line 787, in _parse
    ob.set_data(*args)
  File "Z:\Python\envs\wjahr\playground\lib\site-packages\vispy\gloo\glir.py", line 1639, in set_data
    glTexSubImage3D(self._target, 0, x, y, z, format, gtype, data)
  File "Z:\Python\envs\wjahr\playground\lib\site-packages\vispy\gloo\glir.py", line 1607, in glTexSubImage3D
    width, height, depth, format, type, pixels)
  File "Z:\Python\envs\wjahr\playground\lib\site-packages\OpenGL\latebind.py", line 45, in __call__
    return self._finalCall( *args, **named )
  File "Z:\Python\envs\wjahr\playground\lib\site-packages\OpenGL\wrapper.py", line 882, in wrapperCall
    result = wrappedOperation( *cArguments )
OSError: exception: access violation reading 0x00000126D1C70000

If you suspect this is an IPython bug, please report it at:
    https://github.com/ipython/ipython/issues
or send an email to the mailing list at ipython-dev@python.org

You can print a more detailed traceback right now with "%tb", or use "%debug"
to interactively debug it.
```

And here the detailed traceback:
```
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
Z:\Python\envs\wjahr\playground\lib\site-packages\OpenGL\latebind.py in __call__(self, *args, **named)
     40             try:
---> 41                 return self._finalCall( *args, **named )
     42             except (TypeError,AttributeError) as err:

TypeError: 'NoneType' object is not callable

During handling of the above exception, another exception occurred:

OSError                                   Traceback (most recent call last)
Z:\Python\envs\wjahr\playground\lib\site-packages\vispy\util\event.py in __call__(self, *args, **kwargs)
    453                     continue
    454
--> 455                 self._invoke_callback(cb, event)
    456                 if event.blocked:
    457                     break

Z:\Python\envs\wjahr\playground\lib\site-packages\vispy\util\event.py in _invoke_callback(self, cb, event)
    473             _handle_exception(self.ignore_callback_errors,
    474                               self.print_callback_errors,
--> 475                               self, cb_event=(cb, event))
    476
    477     def _prepare_event(self, *args, **kwargs):

Z:\Python\envs\wjahr\playground\lib\site-packages\vispy\util\event.py in _invoke_callback(self, cb, event)
    469     def _invoke_callback(self, cb, event):
    470         try:
--> 471             cb(event)
    472         except Exception:
    473             _handle_exception(self.ignore_callback_errors,

Z:\Python\envs\wjahr\playground\lib\site-packages\vispy\scene\canvas.py in on_draw(self, event)
    215         # scheduling of further updates
    216         self._update_pending = False
--> 217         self._draw_scene()
    218
    219     def render(self, region=None, size=None, bgcolor=None, crop=None):

Z:\Python\envs\wjahr\playground\lib\site-packages\vispy\scene\canvas.py in _draw_scene(self, bgcolor)
    264             bgcolor = self._bgcolor
    265         self.context.clear(color=bgcolor, depth=True)
--> 266         self.draw_visual(self.scene)
    267
    268     def draw_visual(self, visual, event=None):

Z:\Python\envs\wjahr\playground\lib\site-packages\vispy\scene\canvas.py in draw_visual(self, visual, event)
    302                         else:
    303                             if hasattr(node, 'draw'):
--> 304                                 node.draw()
    305                                 prof.mark(str(node))
    306                 else:

Z:\Python\envs\wjahr\playground\lib\site-packages\vispy\scene\visuals.py in draw(self)
     97         if self.picking and not self.interactive:
     98             return
---> 99         self._visual_superclass.draw(self)
    100
    101

Z:\Python\envs\wjahr\playground\lib\site-packages\vispy\visuals\visual.py in draw(self)
    441         try:
    442             self._program.draw(self._vshare.draw_mode,
--> 443                                self._vshare.index_buffer)
    444         except Exception:
    445             logger.warning("Error drawing visual %r" % self)

Z:\Python\envs\wjahr\playground\lib\site-packages\vispy\visuals\shaders\program.py in draw(self, *args, **kwargs)
     99         self.build_if_needed()
    100         self.update_variables()
--> 101         Program.draw(self, *args, **kwargs)
    102
    103     def build_if_needed(self):

Z:\Python\envs\wjahr\playground\lib\site-packages\vispy\gloo\program.py in draw(self, mode, indices, check_error)
    531
    532         # Process GLIR commands
--> 533         canvas.context.flush_commands()

Z:\Python\envs\wjahr\playground\lib\site-packages\vispy\gloo\context.py in flush_commands(self, event)
    174                 fbo = 0
    175             self.shared.parser.parse([('CURRENT', 0, fbo)])
--> 176         self.glir.flush(self.shared.parser)
    177
    178     def set_viewport(self, *args):

Z:\Python\envs\wjahr\playground\lib\site-packages\vispy\gloo\glir.py in flush(self, parser)
    570         """ Flush all current commands to the GLIR interpreter.
    571         """
--> 572         self._shared.flush(parser)
    573
    574

Z:\Python\envs\wjahr\playground\lib\site-packages\vispy\gloo\glir.py in flush(self, parser)
    492             show = self._verbose if isinstance(self._verbose, str) else None
    493             self.show(show)
--> 494         parser.parse(self._filter(self.clear(), parser))
    495
    496     def _filter(self, commands, parser):

Z:\Python\envs\wjahr\playground\lib\site-packages\vispy\gloo\glir.py in parse(self, commands)
    817
    818         for command in commands:
--> 819             self._parse(command)
    820
    821     def get_object(self, id_):

Z:\Python\envs\wjahr\playground\lib\site-packages\vispy\gloo\glir.py in _parse(self, command)
    785                 ob.set_attribute(*args)
    786             elif cmd == 'DATA':  # VertexBuffer, IndexBuffer, Texture, Shader
--> 787                 ob.set_data(*args)
    788             elif cmd == 'SIZE':  # VertexBuffer, IndexBuffer,
    789                 ob.set_size(*args)  # Texture[1D, 2D, 3D], RenderBuffer

Z:\Python\envs\wjahr\playground\lib\site-packages\vispy\gloo\glir.py in set_data(self, offset, data)
   1637             gl.glPixelStorei(gl.GL_UNPACK_ALIGNMENT, alignment)
   1638         # Upload
-> 1639         glTexSubImage3D(self._target, 0, x, y, z, format, gtype, data)
   1640         # Set alignment back
   1641         if alignment != 4:

Z:\Python\envs\wjahr\playground\lib\site-packages\vispy\gloo\glir.py in glTexSubImage3D(target, level, xoffset, yoffset, zoffset, format, type, pixels)
   1605     depth, height, width = pixels.shape[:3]
   1606     _gl.glTexSubImage3D(target, level, xoffset, yoffset, zoffset,
-> 1607                         width, height, depth, format, type, pixels)
   1608
   1609

Z:\Python\envs\wjahr\playground\lib\site-packages\OpenGL\latebind.py in __call__(self, *args, **named)
     43                 if self._finalCall is None:
     44                     self._finalCall = self.finalise()
---> 45                 return self._finalCall( *args, **named )
     46 if Curry is None:
     47     class Curry(object):

Z:\Python\envs\wjahr\playground\lib\site-packages\OpenGL\wrapper.py in wrapperCall(*args)
    880                                 cArguments = pyArgs
    881                                 try:
--> 882                                     result = wrappedOperation( *cArguments )
    883                                 except ctypes.ArgumentError as err:
    884                                     err.args = err.args + (cArguments,)

OSError: exception: access violation reading 0x00000126D1C70000
```

dataset can be found here:
https://seafile.ist.ac.at/f/e173a2b7c65f4127bd10/

**Comment by sofroniewn** [comment]
Hi @wiebkejahr sorry you're having this problem. You might want to upgrade to `0.2.7`, we just released it yesterday, and there's no immediate reason for me to think that will fix anything here, but you might as well be working off the latest.

Can you tell me the shape and dtype of your data. I also noticed you're using `dask_image` to do the reading, which is fine, but I'm curious if that means your dataset is very large and you want lazy loading, or if there was another reason you went with `dask_image` there.

The ideal case would be that we could reproduce this error with data that is of the form `np.random.random(shape)` or similar using the dask random array

**Comment by wiebkejahr** [comment]
Thanks, totally missed the update, I love the volume rendering functionality already :)
I'll continue testing.

shape: 140x2048x2048
uint16

Why dask_image? At the moment, my data is small enough to load as is (order of 1 GB, but only 1 timepoint). I'm in the process of developing the whole workflow (including different hardware structure available), so trying to stay future proof, and that's why I'm also testing with dask_image.

**Comment by sofroniewn** [comment]
Hmm, `2048` may be the max texture size per axis that your graphics card can handle and this PR #788 introduced code to fix prevent larger textures from being sent over to the graphics card. The limit in the PR is ` > 2048` so the code to do the downsampling won't get triggered in this case.

My two questions are, can you reproduce the problem with `np.random.randint((140, 2048, 2048).astype(np.uint8)` (or similar) and then what happens if you try `np.random.randint((140, 2040, 2040).astype(np.uint8)`, where you just back away slightly from that 2048 number

**Comment by lennart-g** [comment]
Hi, not sure if this belongs here. When I switch to 3D view and start rotating the mesh in the center I get the error below 100% of the time (but only if I've switched to a different slide before or switched to a different one and then switched back to first one).

I'm on windows 8.1 64 bit with NVIDIA geforce gtx 960. Updated my GPU driver today (in an attempt to stop getting the crash) so that's not the issue. When I run the same code on my old lenovo x230 tablet running ubuntu 18.04 64 bit it all works fine.

Would be cool to see this fixed. Learned about napari today and would be exactly what I need ... if it ran :/

Code:
```import napari
import numpy as np
from skimage import data, filters

if __name__ == '__main__':
    blobs_raw = np.stack([
        data.binary_blobs(length=256, n_dim=3, volume_fraction=f)
        for f in np.linspace(0.05, 0.5, 10)
    ])
    blobs = filters.gaussian(blobs_raw, sigma=(0, 2, 2, 2))
    with napari.gui_qt():
        # create a Viewer and add an image here
        viewer = napari.view_image(blobs)
```

Error:

> WARNING: Error drawing visual <vispy.visuals.mesh.MeshVisual object at 0x000000DA3C0331C8>
> WARNING: Traceback (most recent call last):
>   File "C:\Users\lennart\Anaconda3\envs\malaria_seqmentation\lib\site-packages\vispy\app\backends\_qt.py", line 447, in mousePressEvent
>     modifiers=self._modifiers(ev),
>   File "C:\Users\lennart\Anaconda3\envs\malaria_seqmentation\lib\site-packages\vispy\app\base.py", line 181, in _vispy_mouse_press
>     ev = self._vispy_canvas.events.mouse_press(**kwargs)
>   File "C:\Users\lennart\Anaconda3\envs\malaria_seqmentation\lib\site-packages\vispy\util\event.py", line 455, in __call__
>     self._invoke_callback(cb, event)
>   File "C:\Users\lennart\Anaconda3\envs\malaria_seqmentation\lib\site-packages\vispy\util\event.py", line 475, in _invoke_callback
>     self, cb_event=(cb, event))
>   File "C:\Users\lennart\Anaconda3\envs\malaria_seqmentation\lib\site-packages\vispy\util\event.py", line 471, in _invoke_callback
>     cb(event)
>   File "C:\Users\lennart\Anaconda3\envs\malaria_seqmentation\lib\site-packages\vispy\scene\canvas.py", line 344, in _process_mouse_event
>     picked = self.visual_at(event.pos)
>   File "C:\Users\lennart\Anaconda3\envs\malaria_seqmentation\lib\site-packages\vispy\scene\canvas.py", line 397, in visual_at
>     id_ = self._render_picking((fbpos[0], fbpos[1], 1, 1))
>   File "C:\Users\lennart\Anaconda3\envs\malaria_seqmentation\lib\site-packages\vispy\scene\canvas.py", line 483, in _render_picking
>     img = self.render(bgcolor=(0, 0, 0, 0), crop=crop)
>   File "C:\Users\lennart\Anaconda3\envs\malaria_seqmentation\lib\site-packages\vispy\scene\canvas.py", line 257, in render
>     self._draw_scene(bgcolor=bgcolor)
>   File "C:\Users\lennart\Anaconda3\envs\malaria_seqmentation\lib\site-packages\vispy\scene\canvas.py", line 266, in _draw_scene
>     self.draw_visual(self.scene)
>   File "C:\Users\lennart\Anaconda3\envs\malaria_seqmentation\lib\site-packages\vispy\scene\canvas.py", line 304, in draw_visual
>     node.draw()
>   File "C:\Users\lennart\Anaconda3\envs\malaria_seqmentation\lib\site-packages\vispy\scene\visuals.py", line 99, in draw
>     self._visual_superclass.draw(self)
>   File "C:\Users\lennart\Anaconda3\envs\malaria_seqmentation\lib\site-packages\vispy\visuals\visual.py", line 595, in draw
>     v.draw()
>   File "C:\Users\lennart\Anaconda3\envs\malaria_seqmentation\lib\site-packages\vispy\visuals\mesh.py", line 519, in draw
>     Visual.draw(self, *args, **kwds)
>   File "C:\Users\lennart\Anaconda3\envs\malaria_seqmentation\lib\site-packages\vispy\visuals\visual.py", line 443, in draw
>     self._vshare.index_buffer)
>   File "C:\Users\lennart\Anaconda3\envs\malaria_seqmentation\lib\site-packages\vispy\visuals\shaders\program.py", line 101, in draw
>     Program.draw(self, *args, **kwargs)
>   File "C:\Users\lennart\Anaconda3\envs\malaria_seqmentation\lib\site-packages\vispy\gloo\program.py", line 533, in draw
>     canvas.context.flush_commands()
>   File "C:\Users\lennart\Anaconda3\envs\malaria_seqmentation\lib\site-packages\vispy\gloo\context.py", line 176, in flush_commands
>     self.glir.flush(self.shared.parser)
>   File "C:\Users\lennart\Anaconda3\envs\malaria_seqmentation\lib\site-packages\vispy\gloo\glir.py", line 572, in flush
>     self._shared.flush(parser)
>   File "C:\Users\lennart\Anaconda3\envs\malaria_seqmentation\lib\site-packages\vispy\gloo\glir.py", line 494, in flush
>     parser.parse(self._filter(self.clear(), parser))
>   File "C:\Users\lennart\Anaconda3\envs\malaria_seqmentation\lib\site-packages\vispy\gloo\glir.py", line 819, in parse
>     self._parse(command)
>   File "C:\Users\lennart\Anaconda3\envs\malaria_seqmentation\lib\site-packages\vispy\gloo\glir.py", line 779, in _parse
>     ob.draw(*args)
>   File "C:\Users\lennart\Anaconda3\envs\malaria_seqmentation\lib\site-packages\vispy\gloo\glir.py", line 1339, in draw
>     gl.glDrawArrays(mode, first, count)
>   File "C:\Users\lennart\Anaconda3\envs\malaria_seqmentation\lib\site-packages\vispy\gloo\gl\_gl2.py", line 415, in glDrawArrays
>     nativefunc(mode, first, count)
> OSError: exception: access violation reading 0x0000000000000000
>

**Comment by lennart-g** [comment]
Code is from https://ilovesymposia.com/2019/10/24/introducing-napari-a-fast-n-dimensional-image-viewer-in-python/ so I'm pretty sure it's supposed to work

**Comment by sofroniewn** [comment]
Yeah, unfortunately it works for me on my mac too (I just tried it again) - we've had some issues with windows - but we can try chasing this down. Sorry about that!!

There are some other threads on the napari repo which talk about windows issues more explicitly and you might find something in one of those in the meantime. Maybe @tlambert03 knows which ones I have in mind or we can ping some more of our windows users. It comes up so frequently now we should really have a "known windows issues section". Need to work on that ...

**Comment by sofroniewn** [comment]
@lennart-g and @wiebkejahr we made some good progress with our windows specific 3D volume crashes - see a similar issue in #1347 a fix by #1445.

Can both of you try upgrading to napari 0.3.6 and see if the issue still persists? We do still have a couple of lingering ones like #1224 so would be good to know where this one lands. I hope it is now solved!!

**Comment by sofroniewn** [comment]
Given progress on our 3D rendering described above and the lack of activity on this issue I am going to close. If people encounter new 3D rendering issues feel free to reopen or create a new issue
